dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackerOwnershipManager.java

```diff
@@ -254,15 +253,21 @@ public boolean hasAccess( User user, TrackedEntityInstance entityInstance, Progr
             return true;
         }
 
-        OrganisationUnit ou = getOwner( entityInstance.getId(), program, entityInstance::getOrganisationUnit );
+        OrganisationUnit ownerOrgUnit = getOwner( entityInstance.getId(), program );
+        boolean hasOwner = ownerOrgUnit != null;
 
-        if ( program.isOpen() || program.isAudited() )
+        if ( !hasOwner )
         {
-            return organisationUnitService.isInUserSearchHierarchyCached( user, ou );
+            ownerOrgUnit = entityInstance.getOrganisationUnit();
```

Comment:
I would inline `entityInstance.getOrganisationUnit()`. Assigning it to `ownerOrgUnit`  is confusing to me since the program has no owner.